<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Pre-processing input data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/journal.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #303030; color: #cccccc; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #cccccc; background-color: #303030; }
code > span.kw { color: #f0dfaf; } /* Keyword */
code > span.dt { color: #dfdfbf; } /* DataType */
code > span.dv { color: #dcdccc; } /* DecVal */
code > span.bn { color: #dca3a3; } /* BaseN */
code > span.fl { color: #c0bed1; } /* Float */
code > span.ch { color: #dca3a3; } /* Char */
code > span.st { color: #cc9393; } /* String */
code > span.co { color: #7f9f7f; } /* Comment */
code > span.ot { color: #efef8f; } /* Other */
code > span.al { color: #ffcfaf; } /* Alert */
code > span.fu { color: #efef8f; } /* Function */
code > span.er { color: #c3bf9f; } /* Error */
code > span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */
code > span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code > span.sc { color: #dca3a3; } /* SpecialChar */
code > span.vs { color: #cc9393; } /* VerbatimString */
code > span.ss { color: #cc9393; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #f0dfaf; } /* ControlFlow */
code > span.op { color: #f0efd0; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code > span.at { } /* Attribute */
code > span.do { color: #7f9f7f; } /* Documentation */
code > span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code > span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code > span.in { color: #7f9f7f; font-weight: bold; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">About CLIMB</a>
</li>
<li>
  <a href="preprocessing.html">Pre-processing input data</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Four-step procedure
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pairwise_fitting.html">Pairwise fitting</a>
    </li>
    <li>
      <a href="candidate_latent_classes.html">Finding candidate classes</a>
    </li>
    <li>
      <a href="priors.html">Computing prior hyperparameters</a>
    </li>
    <li>
      <a href="running_mcmc.html">Running the MCMC</a>
    </li>
  </ul>
</li>
<li>
  <a href="downstream.html">Downstream analysis</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="https://github.com/hillarykoch/climb_page">
    <span class="fas fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Pre-processing input data</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2021-04-17
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 4 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 3
</p>
<p>
<strong>Knit directory:</strong> <code>lieb/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges" class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of the R Markdown file created these results, you'll want to first commit it to the Git repo. If you're still working on the analysis, you can ignore this warning. When you're finished, you can run <code>wflow_publish</code> to commit the R Markdown file and build the HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it's best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20190717code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20190717)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20190717code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20190717)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongunavailable"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>Session information:</strong> unavailable </a>
</p>
</div>
<div id="strongSessioninformationstrongunavailable" class="panel-collapse collapse">
<div class="panel-body">
<p>Recording the operating system, R version, and package versions is critical for reproducibility. To record the session information, add <code>sessioninfo: &quot;sessionInfo()&quot;</code> to _workflowr.yml. Alternatively, you could use <code>devtools::session_info()</code> or <code>sessioninfo::session_info()</code>. Lastly, you can manually add a code chunk to this file to run any one of these commands and then disable to automatic insertion by changing the workflowr setting to <code>sessioninfo: &quot;&quot;</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongdetected"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>Cache:</strong> detected </a>
</p>
</div>
<div id="strongCachestrongdetected" class="panel-collapse collapse">
<div class="panel-body">
The following chunks had caches available:
<ul>
<li>
unnamed-chunk-11
</li>
</ul>
<p>To ensure reproducibility of the results, delete the cache directory <code>preprocessing_cache</code> and re-run the analysis. To have workflowr automatically delete the cache directory prior to building the file, set <code>delete_cache = TRUE</code> when running <code>wflow_build()</code> or <code>wflow_publish()</code>.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomhillarykochclimbpagetreea49305692caaee2009dd7a906122305505d106a5targetblanka493056a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/hillarykoch/climb_page/tree/a49305692caaee2009dd7a906122305505d106a5" target="_blank">a493056</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomhillarykochclimbpagetreea49305692caaee2009dd7a906122305505d106a5targetblanka493056a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/hillarykoch/climb_page/tree/a49305692caaee2009dd7a906122305505d106a5" target="_blank">a493056</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    analysis/.DS_Store
    Ignored:    analysis/.Rhistory
    Ignored:    analysis/figure/
    Ignored:    analysis/pairwise_fitting_cache/
    Ignored:    analysis/preprocessing_cache/
    Ignored:    analysis/running_mcmc_cache/
    Ignored:    data/.DS_Store
    Ignored:    data/.Rhistory
    Ignored:    output/.DS_Store

Unstaged changes:
    Modified:   analysis/candidate_latent_classes.Rmd
    Modified:   analysis/downstream.Rmd
    Modified:   analysis/preprocessing.Rmd
    Modified:   analysis/priors.Rmd
    Modified:   analysis/running_mcmc.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/preprocessing.Rmd</code>) and HTML (<code>docs/preprocessing.html</code>) files. If you've configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/a49305692caaee2009dd7a906122305505d106a5/analysis/preprocessing.Rmd" target="_blank">a493056</a>
</td>
<td>
hillarykoch
</td>
<td>
2021-04-17
</td>
<td>
update site to match gigantum submission
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hillarykoch/climb_page/dd0b829748283df1a9a8482da76e31744ba504b8/docs/preprocessing.html" target="_blank">dd0b829</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-10-15
</td>
<td>
replace cache
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/ed24c1bc4a5e0542bed7f286a2119436be190df0/analysis/preprocessing.Rmd" target="_blank">ed24c1b</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-08
</td>
<td>
update toc
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hillarykoch/climb_page/ed24c1bc4a5e0542bed7f286a2119436be190df0/docs/preprocessing.html" target="_blank">ed24c1b</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-08
</td>
<td>
update toc
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/194207230fb39f8d5e47c1bb44a77d0c57c1365f/analysis/preprocessing.Rmd" target="_blank">1942072</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-03
</td>
<td>
fix plots on preprocessing slide
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hillarykoch/climb_page/194207230fb39f8d5e47c1bb44a77d0c57c1365f/docs/preprocessing.html" target="_blank">1942072</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-03
</td>
<td>
fix plots on preprocessing slide
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/analysis/preprocessing.Rmd" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
<td>
update preprocessing with quantile transformation for RNA-seq
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hillarykoch/climb_page/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/docs/preprocessing.html" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
<td>
update preprocessing with quantile transformation for RNA-seq
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/cc98d2b3469d9dd1fadf2ba3b9577cd6f35327fd/analysis/preprocessing.Rmd" target="_blank">cc98d2b</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-24
</td>
<td>
replace cache and update to describe flex_mu
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hillarykoch/climb_page/cc98d2b3469d9dd1fadf2ba3b9577cd6f35327fd/docs/preprocessing.html" target="_blank">cc98d2b</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-24
</td>
<td>
replace cache and update to describe flex_mu
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/717e7627599663c41b14ce8f699853d839bde713/analysis/preprocessing.Rmd" target="_blank">717e762</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-23
</td>
<td>
add placeholder for downstream analyses
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hillarykoch/climb_page/717e7627599663c41b14ce8f699853d839bde713/docs/preprocessing.html" target="_blank">717e762</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-23
</td>
<td>
add placeholder for downstream analyses
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/828e725ac9ae66949f0c7821028faa80ae0242c5/analysis/preprocessing.Rmd" target="_blank">828e725</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-23
</td>
<td>
add a preprocessing section
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/hillarykoch/climb_page/828e725ac9ae66949f0c7821028faa80ae0242c5/docs/preprocessing.html" target="_blank">828e725</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-23
</td>
<td>
add a preprocessing section
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="preparing-data-for-input-into-climb" class="section level3">
<h3>Preparing data for input into CLIMB</h3>
<p>CLIMB can, in some cases, be sensitive to data preparation and normalization. This has to do with the relationship between the association labels (<span class="math inline">\(-1, 0, 1\)</span>) and associated parameters. Most notably, if a cluster has a positive (negative) association in dimension <span class="math inline">\(d\)</span>, then the cluster mean will be positive (negative) in dimension <span class="math inline">\(d\)</span> as well. Similarly, if a cluster has null association in dimension <span class="math inline">\(d\)</span>, then the cluster mean will be 0 in dimension <span class="math inline">\(d\)</span>. What this means is that, for data which may have positive, negative, and null associations, we want the data appropriately pre-processed such that these relationships with cluster means will reasonably hold. One should be able to test the suitability of the pre-processing method chosen using only the first step of CLIMB, the pairwise fitting procedure. We will demonstrate what we mean by that later on this page.</p>
<p>In some cases, such as input from a peak-calling routine, preparation of data is straightforward. In others, such has right-skewed data like RNA-seq, preparation may be slightly less so. Here, we walk through steps used for the data analyses in the CLIMB manuscript, plus additional alternatives. These pre-processing steps are:</p>
<ol style="list-style-type: decimal">
<li><a href="#preparation-of-chip-seq-data">Quantile transformation and imputation for peak data</a></li>
<li><a href="#by-log-transformation">Log transformation and quantile normalization of RNA-seq</a></li>
<li><a href="#by-quantile-transformation">Quantile transformation of RNA-seq</a></li>
</ol>
</div>
<div id="preparation-of-chip-seq-data" class="section level3">
<h3>Preparation of ChIP-seq data</h3>
<p>Preparation of data which come from a peak-calling routine is fairly natural. This is because the data are reported as <span class="math inline">\(p\)</span>-values, where small values provide support for the presence of a peak, and the lack of a peak corresponds to &quot;null&quot; behavior. We walk through the data preparation used on the ChIP-seq data in the CLIMB manuscript, though we expect this approach to be more broadly applicable to data output from a peak-calling tool such as MACS2.</p>
<p>First, we assume the user has chosed some peak-calling procedure, with output reported in a format such as ENCODE narrowPeak format, which contains information about <strong>the chromosome, start location, stop location, and <span class="math inline">\(-\log_{10}(p\)</span>-</strong>value) <strong>for the peak</strong>. Next, since the analysis will be across multiple conditions, one needs to align the peaks across the various conditions being analyzed. In our case, we merged overlapping peak locations with <code>bedtools merge</code>. For locations which had a peak in one condition, but lacked a peak in another, we used a <span class="math inline">\(-\log_{10}(p\)</span>-value) of 0 as a placeholder (this correponds to the not-statistically-significant <span class="math inline">\(p\)</span>-value of 1).</p>
<p>As an example, let's read in a 2-dimensional ChIP-seq dataset, where the peak regions have already been aligned with <code>bedtools</code>. We can see that this dataset consists of the typical BED format information (chromosome, start, stop), as well as the <span class="math inline">\(-\log_{10}(p\)</span>-values) for 2 conditions. We see that at row 3, there is a peak in condition 1 that is absent in condition 2, and the reverse is the case at row 6.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pairwise_chip &lt;-<span class="st"> </span>readr::<span class="kw">read_tsv</span>(<span class="st">&quot;data/toy_chip.tsv&quot;</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>)
<span class="kw">head</span>(pairwise_chip)</code></pre></div>
<pre><code># A tibble: 6 x 5
  X1           X2        X3     X4     X5
  &lt;chr&gt;     &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 chr11 100004457 100005627  1.72  22.9  
2 chr11 100006220 100007347  3.06  16.4  
3 chr11 100041141 100041500 10.4    0    
4 chr11 100058679 100059290  0.348  0.719
5 chr11 100168530 100170342 11.5   27.3  
6 chr11 100199832 100201021  0      9.40 </code></pre>
<p>Now, we need to transform these data to input for CLIMB. This is shown below, but involves only a few main steps, as described in the CLIMB manuscript.</p>
<ol style="list-style-type: decimal">
<li>Transform everything to <span class="math inline">\(p\)</span>-values.</li>
<li>Transform <span class="math inline">\(p\)</span>-values to <span class="math inline">\(z\)</span>-scores with the standard normal quantile function</li>
<li>Impute <span class="math inline">\(z\)</span>-scores at locations where a peak was absent in a condition.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)

<span class="co"># BED file information, store this for downstream analysis, e.g. examining loci with GREAT</span>
bed_df &lt;-<span class="st"> </span>dplyr::<span class="kw">select</span>(pairwise_chip, X1, X2, X3)

<span class="co"># Actual data to prepare now</span>
pw_df &lt;-<span class="st"> </span>dplyr::<span class="kw">select</span>(pairwise_chip, X4, X5) %&gt;%
<span class="st">  </span><span class="co"># transform to p-values</span>
<span class="st">  </span>purrr::<span class="kw">map_df</span>(~<span class="st"> </span><span class="dv">10</span> ^<span class="st"> </span>-(.x))

<span class="co"># Filter out rows which have no peaks in any condition</span>
<span class="co">#   (this may happen as a by-product of data normalization, or from data subsetting)</span>
keep_idx &lt;-<span class="st"> </span><span class="kw">apply</span>(pw_df, <span class="dv">1</span>, function(row) !<span class="kw">all</span>(row ==<span class="st"> </span><span class="dv">1</span>))
  
bed_df &lt;-<span class="st"> </span>dplyr::<span class="kw">filter</span>(bed_df, keep_idx)
pw_df &lt;-<span class="st"> </span>dplyr::<span class="kw">filter</span>(pw_df, keep_idx)

<span class="co"># Identify locations with p-value = 1 (these will be imputed)</span>
one_idx &lt;-<span class="st"> </span>pw_df ==<span class="st"> </span><span class="dv">1</span>
  
<span class="co"># Transform data to z-scores using the standard normal quantile function</span>
z &lt;-<span class="st"> </span><span class="kw">qnorm</span>(<span class="kw">as.matrix</span>(pw_df))

<span class="co"># Address numerical overflow</span>
maxval &lt;-<span class="st"> </span><span class="kw">max</span>(z[z !=<span class="st"> </span><span class="ot">Inf</span>])
minval &lt;-<span class="st"> </span><span class="kw">min</span>(z[z !=<span class="st"> </span>-<span class="ot">Inf</span>])
z[z ==<span class="st"> </span><span class="ot">Inf</span>] &lt;-<span class="st"> </span>maxval
z[z ==<span class="st"> </span>-<span class="ot">Inf</span>] &lt;-<span class="st"> </span>minval

<span class="co"># Impute z-scores at loci without peaks</span>
z[one_idx] &lt;-<span class="st"> </span><span class="kw">rnorm</span>(<span class="kw">sum</span>(one_idx))

<span class="co"># Take the negative, so that small p-value ==&gt; large signal</span>
z &lt;-<span class="st"> </span>-z

<span class="co"># Quickly plot the data</span>
<span class="kw">plot</span>(z, <span class="dt">xlim =</span> <span class="kw">c</span>(-<span class="dv">3</span>,<span class="dv">25</span>), <span class="dt">ylim =</span> <span class="kw">c</span>(-<span class="dv">3</span>,<span class="dv">25</span>), <span class="dt">pch =</span> <span class="dv">4</span>, <span class="dt">xlab =</span> <span class="st">&quot;condition 1&quot;</span>, <span class="dt">ylab =</span> <span class="st">&quot;condition 2&quot;</span>)</code></pre></div>
<p><img src="figure/preprocessing.Rmd/unnamed-chunk-3-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/docs/figure/preprocessing.Rmd/unnamed-chunk-3-1.png" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
</tr>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/828e725ac9ae66949f0c7821028faa80ae0242c5/docs/figure/preprocessing.Rmd/unnamed-chunk-3-1.png" target="_blank">828e725</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-23
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that it is critical here that we filtered out the same rows in bed_df (the BED information) and pw_df (the data). This is because, for downstream analyses, we need to track which loci were analyzed.</p>
</div>
<div id="preparation-of-rna-seq-data" class="section level3">
<h3>Preparation of RNA-seq data</h3>
<div id="by-log-transformation" class="section level4">
<h4>By log transformation</h4>
<p>Here, we discuss 2 very similar ways of preparing RNA-seq data for input to CLIMB. First, lets pull some gene expression data from Gene Expression Omnibus. These data concern gene expression in Zebrafish.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(curl)

<span class="co"># Download, unzip, and read in RNA-seq data from GEO</span>
curl::<span class="kw">curl_download</span>(<span class="dt">url =</span> <span class="st">&quot;ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE133nnn/GSE133459/suppl/GSE133459_RSEM_GSE133459_GFP_GSE133459_NFYADN_GSE133459_PBCAB_12hpf_zebrafish_genes_expression_tpm.tsv.gz&quot;</span>,
                    <span class="dt">destfile =</span> <span class="st">&quot;data/tpm_zebrafish.tsv.gz&quot;</span>)
<span class="kw">system</span>(<span class="st">&quot;gunzip data/tpm_zebrafish.tsv.gz&quot;</span>)
TPMs &lt;-<span class="st"> </span>readr::<span class="kw">read_tsv</span>(<span class="st">&quot;data/tpm_zebrafish.tsv&quot;</span>)

<span class="kw">head</span>(TPMs)</code></pre></div>
<pre><code># A tibble: 6 x 11
  gene     transcript  GFP1  GFP2  GFP3 NFYA_DN1 NFYA_DN2 NFYA_DN3 PBCAB1 PBCAB2
  &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 mal2     NM_001135…  2.04  1.39  2.64     3.65     2.21     3.3    3.8    1.6 
2 ftr54    NM_001099…  0     0     0        0        0        0      0      0   
3 foxb1b   NM_131283  24.6  23.6  29.5     26.7     25.9     24.4   36.1   27.6 
4 zgc:174… NM_001113…  4.09  6.09  4.48     7.34     5.74    10.5   10.1    8.54
5 trafd1   NM_001089… 12.6  13.4  14.9     12.1     10.5      9.57  13.2   14.7 
6 si:dkey… NM_001044… 11.3  10.8  17.0      8.61     7.46     8.59   7.03  11.7 
# … with 1 more variable: PBCAB3 &lt;dbl&gt;</code></pre>
<p>We follow some fairly typical normalization strategies to prepare the data for input into CLIMB. Only step 4 is unusual.</p>
<ol style="list-style-type: decimal">
<li>Pool replicates.</li>
<li>Apply a <span class="math inline">\(\log2\)</span> transform, using 1 as an offset to avoid infinities.</li>
<li>Quantile normalize the data.</li>
<li>Apply a location shift to the data such that the positive mode sits over 0.</li>
</ol>
<p>We apply step 4 so that the transformed data better adhere to CLIMB's modeling assumptions. As it is only a location shift, it does not affect the associations between gene expression levels across conditions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(limma)
<span class="kw">library</span>(modeest)

<span class="co">#------------------------------------------------------------------------------</span>
<span class="co"># Pool together replicated data</span>

<span class="co"># First, find unique conditions (here, there are three)</span>
conditions &lt;-<span class="st"> </span>stringr::<span class="kw">str_split</span>(<span class="kw">names</span>(TPMs)[-(<span class="dv">1</span>:<span class="dv">2</span>)], <span class="dt">pattern =</span> <span class="st">&quot;[[:digit:]]&quot;</span>) %&gt;%
<span class="st">  </span>purrr::<span class="kw">map_chr</span>(<span class="dv">1</span>) %&gt;%
<span class="st">  </span>unique

<span class="co"># Pool together TPMs across replicates</span>
merge_dat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">nrow =</span> <span class="kw">nrow</span>(TPMs), <span class="dt">ncol =</span> <span class="kw">length</span>(conditions))
for(i in <span class="kw">seq_along</span>(conditions)) {
    idx &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">pattern =</span> conditions[i], <span class="dt">x =</span> <span class="kw">names</span>(TPMs))
    if(<span class="kw">sum</span>(idx) &gt;<span class="st"> </span><span class="dv">1</span>) {
        merge_dat[,i] &lt;-<span class="st"> </span><span class="kw">rowSums</span>(TPMs[,idx])    
    } else {
        merge_dat[,i] &lt;-<span class="st"> </span>TPMs[,idx]
    }
}
<span class="kw">colnames</span>(merge_dat) &lt;-<span class="st"> </span>conditions

<span class="co">#------------------------------------------------------------------------------</span>
<span class="co"># Do the adjusted quantile normalization</span>

<span class="co"># Remove genes which are 0s in all conditions</span>
keep_idx &lt;-<span class="st"> </span><span class="kw">apply</span>(merge_dat, <span class="dv">1</span>, function(row) !(<span class="kw">all</span>(row ==<span class="st"> </span><span class="dv">0</span>)))
genes &lt;-<span class="st"> </span>TPMs$gene[keep_idx]
merge_dat &lt;-<span class="st"> </span>merge_dat[keep_idx,]

<span class="co"># log2 transform and quantile normalize with limma</span>
l_dat &lt;-<span class="st"> </span><span class="kw">log2</span>(merge_dat +<span class="st"> </span><span class="dv">1</span>)
ql_dat &lt;-<span class="st"> </span>limma::<span class="kw">normalizeQuantiles</span>(l_dat)

<span class="co"># Estimate the central mode</span>
<span class="co">#   It should be the same in every condition, due to quantile normalization</span>
estimated_mode &lt;-<span class="st"> </span>modeest::<span class="kw">mlv</span>(<span class="kw">as.vector</span>(ql_dat), <span class="dt">method =</span> <span class="st">&quot;meanshift&quot;</span>, <span class="dt">par =</span> <span class="dv">5</span>)

<span class="co"># Center the mode over 0</span>
z &lt;-<span class="st"> </span>ql_dat -<span class="st"> </span>estimated_mode
<span class="kw">colnames</span>(z) &lt;-<span class="st"> </span><span class="kw">colnames</span>(merge_dat)

<span class="co"># Make a quick histogram of normalized expression from one condition</span>
<span class="kw">hist</span>(z[,<span class="dv">1</span>], <span class="dt">breaks =</span> <span class="dv">30</span>, <span class="dt">xlab =</span> <span class="st">&quot;normalized transcripts&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Histogram of GFP&quot;</span>)</code></pre></div>
<p><img src="figure/preprocessing.Rmd/unnamed-chunk-5-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-5-1">
Past versions of unnamed-chunk-5-1.png
</button>
</p>
<div id="fig-unnamed-chunk-5-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/docs/figure/preprocessing.Rmd/unnamed-chunk-5-1.png" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>As with the ChIP-seq, we filter out genes which are removed from the analysis, so that we know which genes we examined.</p>
<p>While this data pre-processing strategy was suitable for the analyses in the CLIMB manuscript, it may not always be suitable for all RNA-seq. Here, we demonstrate another variation on this data pre-processing strategy. Here, we will pool single-cell experiments of hematopoietic progenitor cells, take from Olsson et al. 2016 (accession code GSE70245).</p>
<p>After reading in the data, we see that there are 382 cells in total in the experiment, coming from 4 unique cell populations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sc_dat &lt;-<span class="st"> </span>readr::<span class="kw">read_tsv</span>(<span class="st">&quot;data/olsson2016.tsv&quot;</span>)
<span class="kw">head</span>(sc_dat)</code></pre></div>
<pre><code># A tibble: 6 x 383
  uid    Lsk.19 Lsk.20 Lsk.27 Lsk.34 Lsk.36 Lsk.37 Lsk.50 Lsk.57 Lsk.58 Lsk.59
  &lt;chr&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 Marc1       0   0      0      0      0     0          0      0      0   0   
2 March1      0   0      0      0      0     0          0      0      0   0   
3 Marc2       0   7.43   3.28   7.03   2.66  0.934      0      0      0   4.60
4 March2      0   0      0      0      0     1.16       0      0      0   3.67
5 March3      0   0      0      0      3.11  0          0      0      0   0   
6 March4      0   0      0      0      0     0.475      0      0      0   0   
# … with 372 more variables: Lsk.60 &lt;dbl&gt;, Lsk.61 &lt;dbl&gt;, Lsk.62 &lt;dbl&gt;,
#   Lsk.63 &lt;dbl&gt;, Lsk.64 &lt;dbl&gt;, Lsk.65 &lt;dbl&gt;, Lsk.66 &lt;dbl&gt;, Lsk.67 &lt;dbl&gt;,
#   Lsk.69 &lt;dbl&gt;, Lsk.70 &lt;dbl&gt;, Lsk.71 &lt;dbl&gt;, Lsk.72 &lt;dbl&gt;, Lsk.73 &lt;dbl&gt;,
#   Lsk.74 &lt;dbl&gt;, Lsk.75 &lt;dbl&gt;, Lsk.76 &lt;dbl&gt;, Lsk.77 &lt;dbl&gt;, Lsk.78 &lt;dbl&gt;,
#   Lsk.79 &lt;dbl&gt;, Lsk.80 &lt;dbl&gt;, Lsk.81 &lt;dbl&gt;, Lsk.82 &lt;dbl&gt;, Lsk.83 &lt;dbl&gt;,
#   Lsk.84 &lt;dbl&gt;, Lsk.85 &lt;dbl&gt;, Lsk.86 &lt;dbl&gt;, Lsk.87 &lt;dbl&gt;, Lsk.88 &lt;dbl&gt;,
#   Lsk.89 &lt;dbl&gt;, Lsk.90 &lt;dbl&gt;, Lsk.91 &lt;dbl&gt;, Lsk.92 &lt;dbl&gt;, Lsk.93 &lt;dbl&gt;,
#   Lsk.94 &lt;dbl&gt;, Lsk.95 &lt;dbl&gt;, Lsk.96 &lt;dbl&gt;, Lsk.10 &lt;dbl&gt;, Lsk.11 &lt;dbl&gt;,
#   Lsk.12 &lt;dbl&gt;, Lsk.13 &lt;dbl&gt;, Lsk.14 &lt;dbl&gt;, Lsk.15 &lt;dbl&gt;, Lsk.16 &lt;dbl&gt;,
#   Lsk.17 &lt;dbl&gt;, Lsk.18 &lt;dbl&gt;, Lsk.1 &lt;dbl&gt;, Lsk.21 &lt;dbl&gt;, Lsk.22 &lt;dbl&gt;,
#   Lsk.23 &lt;dbl&gt;, Lsk.24 &lt;dbl&gt;, Lsk.26 &lt;dbl&gt;, Lsk.28 &lt;dbl&gt;, Lsk.29 &lt;dbl&gt;,
#   Lsk.2 &lt;dbl&gt;, Lsk.30 &lt;dbl&gt;, Lsk.31 &lt;dbl&gt;, Lsk.32 &lt;dbl&gt;, Lsk.33 &lt;dbl&gt;,
#   Lsk.35 &lt;dbl&gt;, Lsk.38 &lt;dbl&gt;, Lsk.39 &lt;dbl&gt;, Lsk.3 &lt;dbl&gt;, Lsk.40 &lt;dbl&gt;,
#   Lsk.41 &lt;dbl&gt;, Lsk.42 &lt;dbl&gt;, Lsk.43 &lt;dbl&gt;, Lsk.44 &lt;dbl&gt;, Lsk.45 &lt;dbl&gt;,
#   Lsk.46 &lt;dbl&gt;, Lsk.47 &lt;dbl&gt;, Lsk.48 &lt;dbl&gt;, Lsk.49 &lt;dbl&gt;, Lsk.4 &lt;dbl&gt;,
#   Lsk.51 &lt;dbl&gt;, Lsk.52 &lt;dbl&gt;, Lsk.53 &lt;dbl&gt;, Lsk.54 &lt;dbl&gt;, Lsk.55 &lt;dbl&gt;,
#   Lsk.5 &lt;dbl&gt;, Lsk.6 &lt;dbl&gt;, Lsk.7 &lt;dbl&gt;, Lsk.8 &lt;dbl&gt;, Lsk.9 &lt;dbl&gt;,
#   Cmp.10 &lt;dbl&gt;, Cmp.11 &lt;dbl&gt;, Cmp.12 &lt;dbl&gt;, Cmp.14 &lt;dbl&gt;, Cmp.15 &lt;dbl&gt;,
#   Cmp.16 &lt;dbl&gt;, Cmp.17 &lt;dbl&gt;, Cmp.18 &lt;dbl&gt;, Cmp.19 &lt;dbl&gt;, Cmp.20 &lt;dbl&gt;,
#   Cmp.21 &lt;dbl&gt;, Cmp.22 &lt;dbl&gt;, Cmp.23 &lt;dbl&gt;, Cmp.24 &lt;dbl&gt;, Cmp.25 &lt;dbl&gt;,
#   Cmp.26 &lt;dbl&gt;, Cmp.27 &lt;dbl&gt;, …</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">uid &lt;-<span class="st"> </span>sc_dat$uid
sc_dat$uid &lt;-<span class="st"> </span><span class="ot">NULL</span>

<span class="co"># Extract cell types</span>
cells &lt;-<span class="st"> </span><span class="kw">str_split</span>(<span class="kw">colnames</span>(sc_dat), <span class="dt">pattern =</span> <span class="st">&quot;</span><span class="ch">\\</span><span class="st">.&quot;</span>) %&gt;%
<span class="st">    </span><span class="kw">map_chr</span>(<span class="dv">1</span>)
<span class="kw">unique</span>(cells)</code></pre></div>
<pre><code>[1] &quot;Lsk&quot; &quot;Cmp&quot; &quot;Gmp&quot; &quot;LK&quot; </code></pre>
<p>Again, we will pool the data by cell type, apply a log2 transformation with offset, and quantile normalize.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#------------------------------------------------------------------------------</span>
<span class="co"># Pool together replicated data</span>
sc_merge_dat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="ot">NA</span>, <span class="dt">nrow =</span> <span class="kw">nrow</span>(sc_dat), <span class="dt">ncol =</span> <span class="kw">length</span>(<span class="kw">unique</span>(cells)))
for(i in <span class="kw">seq_along</span>(<span class="kw">unique</span>(cells))) {
    cell_idx &lt;-<span class="st"> </span>cells ==<span class="st"> </span><span class="kw">unique</span>(cells)[i]
    sc_merge_dat[,i] &lt;-<span class="st"> </span><span class="kw">rowSums</span>(sc_dat[,cell_idx])
}
<span class="kw">colnames</span>(sc_merge_dat) &lt;-<span class="st"> </span><span class="kw">unique</span>(cells)

<span class="co">#------------------------------------------------------------------------------</span>
<span class="co"># Do the adjusted quantile normalization</span>

<span class="co"># Remove genes which are 0s in all conditions</span>
keep_idx &lt;-<span class="st"> </span><span class="kw">apply</span>(sc_merge_dat, <span class="dv">1</span>, function(X) !<span class="kw">all</span>(X ==<span class="st"> </span><span class="dv">0</span>))
sc_merge_dat &lt;-<span class="st"> </span>sc_merge_dat[keep_idx,]
uid &lt;-<span class="st"> </span>uid[keep_idx]

<span class="co"># log2 transform and quantile normalize with limma</span>
l_dat &lt;-<span class="st"> </span><span class="kw">log2</span>(sc_merge_dat +<span class="st"> </span><span class="dv">1</span>)
ql_dat &lt;-<span class="st"> </span>limma::<span class="kw">normalizeQuantiles</span>(l_dat)</code></pre></div>
<p>However, if we plot the histogram of these data, we notice they have a sligtly different distributional structure than the previous Zebrafish data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(ql_dat[,<span class="dv">1</span>], <span class="dt">breaks =</span> <span class="dv">30</span>, <span class="dt">xlab =</span> <span class="st">&quot;normalized transcripts&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Histogram of LK&quot;</span>)
<span class="kw">abline</span>(<span class="dt">v =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="fl">4.5</span>), <span class="dt">col =</span> <span class="st">&quot;darkcyan&quot;</span>)</code></pre></div>
<p><img src="figure/preprocessing.Rmd/unnamed-chunk-8-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-8-1">
Past versions of unnamed-chunk-8-1.png
</button>
</p>
<div id="fig-unnamed-chunk-8-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/docs/figure/preprocessing.Rmd/unnamed-chunk-8-1.png" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
</tr>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/828e725ac9ae66949f0c7821028faa80ae0242c5/docs/figure/preprocessing.Rmd/unnamed-chunk-8-1.png" target="_blank">828e725</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-04-23
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Most importantly, there is a flattening of the disrtibution between the vertical cyan lines. This motivates us to apply a location shift such that this flat region lies over the origin, rather than the positive mode (here, the positive mode is around 7.5). To do this, we approximate the first and second derivitaves of the density using finite differences, and apply a location shift to the data such that the positive inflection point sits over 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#------------------------------------------------------------------------------</span>
<span class="co"># Apply a location shift at the inflection point</span>

<span class="co"># Approximate the density</span>
dens &lt;-<span class="st"> </span><span class="kw">density</span>(<span class="kw">unlist</span>(ql_dat))

<span class="co"># Approximate the first and second derivatives of the density</span>
first_deriv &lt;-<span class="st"> </span><span class="kw">diff</span>(dens$y) /<span class="st"> </span><span class="kw">diff</span>(dens$x)
second_deriv &lt;-<span class="st"> </span><span class="kw">diff</span>(first_deriv) /<span class="st"> </span><span class="kw">diff</span>(<span class="kw">tail</span>(dens$x, -<span class="dv">1</span>))

<span class="co"># Find inflections</span>
flex &lt;-<span class="st"> </span><span class="kw">c</span>()
for(i in <span class="dv">2</span>:<span class="kw">length</span>(second_deriv)){
    if(<span class="kw">sign</span>(second_deriv[i]) !=<span class="st"> </span><span class="kw">sign</span>(second_deriv[i<span class="dv">-1</span>])){
        flex &lt;-<span class="st"> </span><span class="kw">c</span>(flex, i)
    }
}

<span class="co"># Find which inflection point coincides with first derivative being approximately 0</span>
<span class="co">#   Apply the location shift accordingly</span>
flattening &lt;-<span class="st"> </span>dens$x[flex[<span class="kw">which.min</span>(<span class="kw">abs</span>(first_deriv[flex]))]]
z &lt;-<span class="st"> </span>ql_dat -<span class="st"> </span>flattening

<span class="co"># Make a quick histogram and plot of normalized expression from one condition</span>
<span class="kw">hist</span>(z[,<span class="dv">1</span>], <span class="dt">breaks =</span> <span class="dv">30</span>, <span class="dt">xlab =</span> <span class="st">&quot;normalized transcripts&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Histogram of LK, after shifting&quot;</span>)</code></pre></div>
<p><img src="figure/preprocessing.Rmd/unnamed-chunk-9-1.png" width="384" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-9-1">
Past versions of unnamed-chunk-9-1.png
</button>
</p>
<div id="fig-unnamed-chunk-9-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/docs/figure/preprocessing.Rmd/unnamed-chunk-9-1.png" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This transformation may be more widely applicable to RNA-seq datasets.</p>
</div>
<div id="by-quantile-transformation" class="section level4">
<h4>By quantile transformation</h4>
<p>A final, straightforward method to preprocessing RNA-seq data involves quantile transformation, just like with peak-calling data. This method has the advantage of being simple, while also assuring greater suitability of CLIMB's assumption that data labeled 0 in some dimension marginally follow a <span class="math inline">\(Z\)</span>-score distribution (sometimes called the theoretical null). Let's apply this method to the previously shown RNA-seq datasets.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get the normalized ranks (empirical CDF) which are uniformly distributed</span>
<span class="co"># then apply the quantile function of the standard normal</span>

<span class="co"># RECALL: merge_dat is the Zebrafish RNA-seq data</span>
z &lt;-<span class="st"> </span><span class="kw">apply</span>(merge_dat, <span class="dv">2</span>, function(X) <span class="kw">rank</span>(X, <span class="dt">ties.method =</span> <span class="st">&quot;random&quot;</span>) /<span class="st"> </span>(<span class="kw">length</span>(X) +<span class="st"> </span><span class="dv">1</span>)) %&gt;%
<span class="st">  </span>qnorm

<span class="co"># RECALL: sc_merge_dat is the pooled single-cell RNA-seq data</span>
sc_z &lt;-<span class="st"> </span><span class="kw">apply</span>(sc_merge_dat, <span class="dv">2</span>, function(X) <span class="kw">rank</span>(X, <span class="dt">ties.method =</span> <span class="st">&quot;random&quot;</span>) /<span class="st"> </span>(<span class="kw">length</span>(X) +<span class="st"> </span><span class="dv">1</span>)) %&gt;%
<span class="st">  </span>qnorm

<span class="co"># We see from these histograms that the data are standard normal</span>
<span class="kw">hist</span>(z[,<span class="dv">1</span>], <span class="dt">breaks =</span> <span class="dv">35</span>, <span class="dt">xlab =</span> <span class="st">&quot;normalized transcripts&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Histogram of LSK&quot;</span>)
<span class="kw">hist</span>(sc_z[,<span class="dv">1</span>], <span class="dt">breaks =</span> <span class="dv">35</span>, <span class="dt">xlab =</span> <span class="st">&quot;normalized transcripts&quot;</span>, <span class="dt">main =</span> <span class="st">&quot;Histogram of GFP&quot;</span>)

<span class="co"># But the joint distributions look like mixtures</span>
<span class="kw">plot</span>(z[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)], <span class="dt">xlab =</span> <span class="kw">colnames</span>(z)[<span class="dv">2</span>], <span class="dt">ylab =</span> <span class="kw">colnames</span>(z)[<span class="dv">3</span>], <span class="dt">main =</span> <span class="st">&quot;Zebrafish RNA-seq&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)
<span class="kw">plot</span>(sc_z[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>)], <span class="dt">xlab =</span> <span class="kw">colnames</span>(sc_z)[<span class="dv">1</span>], <span class="dt">ylab =</span> <span class="kw">colnames</span>(sc_z)[<span class="dv">4</span>], <span class="dt">main =</span> <span class="st">&quot;Pooled single-cell RNA-seq&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)</code></pre></div>
<p><img src="figure/preprocessing.Rmd/unnamed-chunk-10-1.png" width="1344" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/194207230fb39f8d5e47c1bb44a77d0c57c1365f/docs/figure/preprocessing.Rmd/unnamed-chunk-10-1.png" target="_blank">1942072</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-03
</td>
</tr>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/docs/figure/preprocessing.Rmd/unnamed-chunk-10-1.png" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Using the normalized ranks and standard normal quantile transformation may cause a sort of &quot;compressing&quot; of the data. This is because ranking the data obscured distances between various clusters in the mixture. Then, when applying the typical pairwise fitting step, the cluster estimation may be sub-optimal. To resolve this, we suggest placing a lower bound on the absolute value of the class means for dimensions associated with some signal (i.e., -1 or 1), while allowing more flexibility in the estimation of the positive and negative means. It is important to note that such constraints only exist in the pairwise fitting step, and no longer pose any issue downstream.</p>
<p>Here, we fit a pairwise model to the Zebrafish and pooled single cell datasets. We fit the model under three scenarios:</p>
<ol style="list-style-type: decimal">
<li>A model with no lower bound on the magnitude of the class means, which is also less flexible in estimation of positive and negative associations</li>
<li>A model with no lower bound on the magnitude of the class means, but which is more flexible in estimation of positive and negative associations</li>
<li>A model with a lower bound on the magnitude of the class means, which is also more flexible in estimation of positive and negative associations</li>
</ol>
<p>We plot the data colored by their maximum <em>a posteriori</em> class labels. We can see that <strong>modeling approach 3</strong> appears the most robust, as approach 1 is not flexible enough, and approach 2 encounters identifiability issues, resulting in an unwanted merging of clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))

pal &lt;-<span class="st"> </span>CLIMB::<span class="kw">get_pals</span>(<span class="dv">1</span>)

fit_no_bound &lt;-<span class="st"> </span>CLIMB::<span class="kw">get_pairwise_fits</span>(z[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)], <span class="dt">nlambda =</span> <span class="dv">3</span>, <span class="dt">parallel =</span> <span class="ot">FALSE</span>, <span class="dt">bound =</span> <span class="dv">0</span>)
fit_no_bound_flex &lt;-<span class="st"> </span>CLIMB::<span class="kw">get_pairwise_fits</span>(z[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)], <span class="dt">nlambda =</span> <span class="dv">3</span>, <span class="dt">parallel =</span> <span class="ot">FALSE</span>, <span class="dt">bound =</span> <span class="dv">0</span>, <span class="dt">flex_mu =</span> <span class="ot">TRUE</span>)
fit_bound_flex &lt;-<span class="st"> </span>CLIMB::<span class="kw">get_pairwise_fits</span>(z[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)], <span class="dt">nlambda =</span> <span class="dv">3</span>, <span class="dt">parallel =</span> <span class="ot">FALSE</span>, <span class="dt">bound =</span> <span class="kw">qnorm</span>(.<span class="dv">8</span>), <span class="dt">flex_mu =</span> <span class="ot">TRUE</span>)

sc_fit_no_bound &lt;-<span class="st"> </span>CLIMB::<span class="kw">get_pairwise_fits</span>(sc_z[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>)], <span class="dt">nlambda =</span> <span class="dv">3</span>, <span class="dt">parallel =</span> <span class="ot">FALSE</span>, <span class="dt">bound =</span> <span class="dv">0</span>)
sc_fit_no_bound_flex &lt;-<span class="st"> </span>CLIMB::<span class="kw">get_pairwise_fits</span>(sc_z[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>)], <span class="dt">nlambda =</span> <span class="dv">3</span>, <span class="dt">parallel =</span> <span class="ot">FALSE</span>, <span class="dt">bound =</span> <span class="dv">0</span>, <span class="dt">flex_mu =</span> <span class="ot">TRUE</span>)
sc_fit_bound_flex &lt;-<span class="st"> </span>CLIMB::<span class="kw">get_pairwise_fits</span>(sc_z[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>)], <span class="dt">nlambda =</span> <span class="dv">3</span>, <span class="dt">parallel =</span> <span class="ot">FALSE</span>, <span class="dt">bound =</span> <span class="kw">qnorm</span>(.<span class="dv">8</span>), <span class="dt">flex_mu =</span> <span class="ot">TRUE</span>)

<span class="kw">plot</span>(z[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)], <span class="dt">col =</span> pal[fit_no_bound[[<span class="dv">1</span>]]$cluster], <span class="dt">main =</span> <span class="st">&quot;Approach 1&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)
<span class="kw">plot</span>(z[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)], <span class="dt">col =</span> pal[fit_no_bound_flex[[<span class="dv">1</span>]]$cluster], <span class="dt">main =</span> <span class="st">&quot;Approach 2&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)
<span class="kw">plot</span>(z[,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>)], <span class="dt">col =</span> pal[fit_bound_flex[[<span class="dv">1</span>]]$cluster], <span class="dt">main =</span> <span class="st">&quot;Approach 3&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)

<span class="kw">plot</span>(sc_z[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>)], <span class="dt">col =</span> pal[sc_fit_no_bound[[<span class="dv">1</span>]]$cluster], <span class="dt">main =</span> <span class="st">&quot;Approach 1&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)
<span class="kw">plot</span>(sc_z[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>)], <span class="dt">col =</span> pal[sc_fit_no_bound_flex[[<span class="dv">1</span>]]$cluster], <span class="dt">main =</span> <span class="st">&quot;Approach 2&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)
<span class="kw">plot</span>(sc_z[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">4</span>)], <span class="dt">col =</span> pal[sc_fit_bound_flex[[<span class="dv">1</span>]]$cluster], <span class="dt">main =</span> <span class="st">&quot;Approach 3&quot;</span>, <span class="dt">pch =</span> <span class="dv">4</span>)</code></pre></div>
<p><img src="figure/preprocessing.Rmd/unnamed-chunk-11-1.png" width="1008" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/194207230fb39f8d5e47c1bb44a77d0c57c1365f/docs/figure/preprocessing.Rmd/unnamed-chunk-11-1.png" target="_blank">1942072</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-03
</td>
</tr>
<tr>
<td>
<a href="https://github.com/hillarykoch/climb_page/blob/41ecdb67ed5cfe173b5a641f89d3749ce80e3fbb/docs/figure/preprocessing.Rmd/unnamed-chunk-11-1.png" target="_blank">41ecdb6</a>
</td>
<td>
hillarykoch
</td>
<td>
2020-05-02
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
